File: use_case/interview/create-interview.py
from copy import Error

from model.interview.interview import Interview


class CreateInterview:
    def __init__(self, interview_repository, book_room):
        self.interviewRepository = interview_repository
        self.book_room = book_room

    def create(self, start_date, end_date, interviewer, position, candidate, room):
        if not position.available(candidate.skills):
            raise Error("Position not available for those skills")

        if not room.is_available(start_date, end_date):
            raise Error("Room not available")

        if not interviewer.available(start_date, end_date):
            raise Error("Interviewer not available")

        if not candidate.available(start_date, end_date):
            raise Error("Candidate not available")

        self.book_room(room, start_date, end_date, interviewer, candidate)

        self.interview_repository.save(Interview(
            start_date,
            end_date,
            interviewer,
            position,
            candidate,
            room
        ))

File: src/core/use_case/create-interview.ts
import {InterviewRepositoryInterface} from "../interface/interview-repository.interface";
import {Candidate} from "../business/candidate/candidate.model";
import {RecruiterRepositoryInterface} from "../interface/recruiter-repository.interface";
import {RoomRepositoryInterface} from "../interface/room-repository.interface";
import {Interview} from "../business/interview/interview.model";

export class CreateInterview {
  constructor(
    private readonly interviewRepository: InterviewRepositoryInterface,
    private readonly recruiterRepository: RecruiterRepositoryInterface,
    private readonly roomRepository: RoomRepositoryInterface,
  ) {}

  plan(candidate: Candidate): Interview {
    const recruiters = this.recruiterRepository.getRecruitersBySkills(
      candidate.skills
    );

    if (recruiters.length === 0) {
      throw new Error('No recruiter available for those skills');
    }

    for (const availability of candidate.availabilities) {
      const availableRecruiter = recruiters.find((r) => r.isAvailable(availability[0], availability[1]));

      if (availableRecruiter) {
        const availableRoom = this.roomRepository.getRoomByAvailability(availability);

        if (availableRoom) {
          availableRoom.book(availability);
          this.roomRepository.save(availableRoom);

          return this.interviewRepository.create(
            availability,
            availableRoom,
            candidate,
            availableRecruiter
          );
        }
      }
    }

    throw new Error('No room or recruiter available');
  }
}
